### DISCUSSION TITLE
"Nested" usage of `declared_attr`

### DISCUSSION THREAD

**Discussion by pykenny** [sqlalchemy/sqlalchemy#8914] [Usage Questions] [answered]
A follow-up question from [this post](https://github.com/sqlalchemy/sqlalchemy/discussions/8911). Not sure if it's a bug so finally make it into a discussion thread.

As previously mentioned, I'm defining mixins that define table index setup for some model schemas, which requires usage of [`declared_attr`](https://docs.sqlalchemy.org/en/14/orm/mapping_api.html#sqlalchemy.orm.declared_attr) to generate unique names by using table name as identifier, and then merge the mixins along with other mixins with `declared_attr` again to form up `__table_args__` attribute in the resulting models. Unlike [official example](https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html#creating-indexes-with-mixins) that only lets the inherited class to manage `__table_args__` on its own, managing `__table_args__` through `declared_attr` in the resulting table just doesn't work:

```python
# SA 1.4
from my_orm.base import Base  # <= Declarative ORM base class
from sqlalchemy import Column, Integer, String
from sqlalchemy.schema import Index
from sqlalchemy.orm import declarative_mixin, declared_attr

@declarative_mixin
class Mixin:
    @declared_attr
    def __table_args__(cls):
        return (Index(f"my_index_{cls.__tablename__}", "name"), )

class MyModel(Mixin, Base):
    __tablename__ = "my_table"

    @declared_attr
    def __table_args__(cls):
        return (*Mixin.__table_args__, {"schema": "my_database"})

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)
```

Running the script above leads to this error:
```text
AttributeError: type object 'Mixin' has no attribute '__tablename__'
```

**Comment by zzzeek** [comment, marked as answer]
the stack trace here shows the problem:

```py
 File "/home/classic/dev/sqlalchemy/lib/sqlalchemy/orm/decl_api.py", line 227, in __get__
    return desc.fget(cls)
  File "/home/classic/dev/sqlalchemy/test3.py", line 14, in __table_args__
    return (Index(f"my_index_{cls.__tablename__}", "name"), )
AttributeError: type object 'Mixin' has no attribute '__tablename__'. Did you mean: '__tablename__'?

```

you're calling the `Mixin.__table_args__` accessor, which when using `@declared_attr` is basically equivalent to using `@classmethod`.  the method is then looking upon its class for an attribute named `__tablename__` , which it doesnt have.

This doesnt really have anything to do with SQLAlhcemy, here's your same error using plain python objects:

```py
class Mixin:
    @classmethod
    def __table_args__(cls):
        return f"my_index_{cls.__tablename__}"

class MyModel:
    __tablename__ = "my_table"

    @classmethod
    def __table_args__(cls):
        return Mixin.__table_args__()


print(MyModel.__table_args__())
```

or to illustrate it using an attribute based classmethod

```py
class classproperty(property):
    """A decorator that behaves like @property except that operates
    on classes rather than instances.
    """

    def __init__(self, fget, *arg, **kw):
        super(classproperty, self).__init__(fget, *arg, **kw)
        self.__doc__ = fget.__doc__

    def __get__(desc, self, cls):
        return desc.fget(cls)

class Mixin:
    @classproperty
    def __table_args__(cls):
        return ((f"my_index_{cls.__tablename__}", "name"), )

class MyModel:
    __tablename__ = "my_table"

    @classproperty
    def __table_args__(cls):
        return (*Mixin.__table_args__, {"schema": "my_database"})




print(MyModel.__table_args__)
```


seems like you intended `MyModel` to inherit from `Mixin` and use `super()` to call it in terms of the superclass:

```py
class MyModel(Mixin, Base):
    __tablename__ = "my_table"

    @declared_attr
    def __table_args__(cls):
        return (*super().__table_args__, {"schema": "my_database"})
```

if you needed more specificity than that, you'd need to bypass the `@declared_attr` on `Mixin` and provide methods where you can pass the desired target `cls` explicitly.

**Reply by pykenny** [reply]
PyCharm thinks `cls` refers to some realized instance and keeps asking me to rename it (to `self`), so I thought so as well (`__tablename__` should already been available when `Mixin.__table_args__()` gets called). It makes more sense if it's actually a class method.

**Reply by pykenny** [reply]
Then what if I'm inheriting more than one mixins require access to `__tablename__`? `super()` only resolves naming to the nearest parent in MRO so seems like I'll need to tweak a little bit:
```python
# SA 1.4
from my_orm.base import Base  # <= Declarative ORM base class
from sqlalchemy import Column, Integer, String
from sqlalchemy.schema import Index
from sqlalchemy.orm import declarative_mixin, declared_attr

@declarative_mixin
class MixinOne:
    @declared_attr
    def __table_args__(cls):
        return (Index(f"my_index_{cls.__tablename__}", "name"), )

@declarative_mixin
class MixinTwo:
    @declared_attr
    def __table_args__(cls):
        return (Index(f"my_index2_{cls.__tablename__}", "category"), )

class MyModel(MixinOne, MixinTwo, Base):
    __tablename__ = "my_table"

    @declared_attr
    def __table_args__(cls):
        return (*super().__table_args__, *super(MixinOne, cls).__table_args__, {"schema": "my_database"})

    id = Column(Integer, primary_key=True)
    name = Column(String, nullable=False)
    category = Column(Integer, nullable=False)
```

**Reply by CaselIT** [reply]
You can control what superclass super uses by calling it with two arguments https://docs.python.org/3/library/functions.html?highlight=super#super

**Reply by pykenny** [reply]
I see that. That's how the `*super(MixinOne, cls)` part managed to access `MixinTwo.__table_args__`.
